### PR TITLE
METRON-857 Ability to completely build project in single Docker Image

### PR DESCRIPTION
**Pull request by ottobackwards**
## Contributor Comments

Before ambari integration, we had the ability to use the ansible docker image to build and deploy metron, and a workflow around that.

With the ambari integration and required rpm-docker build stage that was lost.

This PR makes the necessary changes such that metron can be built either in the ansible docker image ( including building the build-rpms profile ) or as currently documented ( only spawning docker when building -P build-rpms ).

To do this, two profiles, with automatic activation where created, each with the correct maven execute task for the environment.  If we are in docker already, we just execute the script, if we are not, we execute docker.

To test this (docker must be running of course ): 
Build the product normally, and build the rpms 

```
$>cd incubator-metron
$>mvn clean install -D skipTests
$>cd metron-deployment
$> mvn clean package -P build-rpms
```

This should work normally. You should see the RPMS user rpm-docker/RPMS/noarch/

Then
```
$>mvn clean
$>cd ..
$>mvn clean
$>cd metron-deployment/packaging/docker/ansible-docker
$>docker build -t ansible-docker:2.0.0.2 .
$>docker run -it -v {PATH-TO-METRON}/incubator-metron/:/root/incubator-metron ansible-docker:2.0.0.2 bash
$docker>cd incubator-metron
$docker>mvn clean install -DskipTests
$docker>cd metron-deployment
$docker>mvn clean package -P build-rpms
```

As before you should see the rpms. 

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron (Incubating).  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [ x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x ] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x ] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x ] Have you ensured that the full suite of tests and checks have been executed in the root incubating-metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  bin/generate-md.sh
  mvn site:site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile [hunk 1]
```diff
@@ -21,10 +21,10 @@ RUN yum install -y tar
 RUN yum install -y wget
 RUN yum groupinstall -y "Development tools"
 RUN yum install -y zlib-dev openssl-devel sqlite-devel bzip2-devel libffi-devel
-RUN wget https://www.python.org/ftp/python/2.7.10/Python-2.7.10.tgz -O /usr/src/Python-2.7.10.tgz
+RUN wget https://www.python.org/ftp/python/2.7.12/Python-2.7.12.tgz -O /usr/src/Python-2.7.12.tgz
```
Comment (JonZeolla) [inline metron-deployment/packaging/docker/ansible-docker/Dockerfile]: I thought we were pinning to 2.7.11 generally.  I know it's a bit pedantic, as the differences are minimal, but perhaps we should just stay with a single version across all of our builds?

File: metron-deployment/packaging/docker/rpm-docker/pom.xml [hunk 2]
```diff
@@ -227,4 +183,97 @@
...
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+            <id>NeedsDocker</id>
+            <activation>
```
Comment (dlyle65535) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: Won't doing this break the Ansible deployment for quick-dev and full-dev? Couldn't you just leave the build-rpms profile and add your 'InDocker' profile?

Comment (ottobackwards) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: This is a layer below that.  The build-rpms is when you build deployment.

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile (hunk 1 above)
Comment (ottobackwards) [inline metron-deployment/packaging/docker/ansible-docker/Dockerfile]: Sure, let me 2x check that

File: metron-deployment/packaging/docker/rpm-docker/pom.xml (hunk 2 above)
Comment (ottobackwards) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: - I think this is the proper level to have this profile
- I prefer two clear profiles to one profile that acts as an override

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile (hunk 1 above)
Comment (ottobackwards) [inline metron-deployment/packaging/docker/ansible-docker/Dockerfile]: It is really hard to find something that say what version of python, unless I'm just not looking correctly.  I did find it in a vagrant deployment doc.  I'll change it.

File: metron-deployment/packaging/docker/rpm-docker/pom.xml (hunk 2 above)
Comment (ottobackwards) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: To clarify the way it works:

At metron-deploymnet:pom.xml => Should we build RPM's?
At rpm-docker:pom.xml => do we need to start docker or just call script

Comment (dlyle65535) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: So, I'm good so long as -Pbuild-rpms works starts a Docker container when called from metron-deployment or the top level directory. So perhaps make the NeedsDocker the default activation at this level? The default behavior should be preserved.

Comment (ottobackwards) [inline metron-deployment/packaging/docker/rpm-docker/pom.xml]: The way I did it with activation, you don't have to do anything.  The only thing you need to do is the -P build-rams as before.  The POM automatically activates the appropriate behavior.  Start docker unless we are in docker.

**Comment by JonZeolla** [comment]
I wasn't able to spin this up without changing the Dockerfile to be `FROM centos:centos6` instead of `FROM centos:6.6`.  This also matches rpm-docker.  Any concern with making that change here?  I don't know why we would want to pin to an older version of centos anyway, especially because centos6 is done releasing.

**Comment by ottobackwards** [comment]
done

**Comment by JonZeolla** [comment]
+1 thanks @ottobackwards

**Pull request closed without merging**